File: assets/Scripts/player/controllers/PlayerCameraController.js
// FILE: Scripts/player/controllers/PlayerCameraController.js
"use strict";

const PlayerCamera = require("../PlayerCamera.js");

function req(v, msg) {
    if (v == null) throw new Error(msg);
    return v;
}

class PlayerCameraController {
    constructor(player) {
        this.player = req(player, "[PlayerCameraController] player is required");
        this.impl = null;
    }

    onStart() {
        const pawn = req(this.player.pawn, "[PlayerCameraController] player.pawn is required");
        this.impl = new PlayerCamera(pawn);
        this.impl.attach();
    }

    onUpdate(dt) {
        const pawn = req(this.player.pawn, "[PlayerCameraController] player.pawn is required");
        const impl = req(this.impl, "[PlayerCameraController] impl is null");

        impl.update(pawn.frame);

        pawn.frame.view.yaw = impl.getYaw();
        pawn.frame.view.pitch = impl.getPitch();
        pawn.frame.view.type = impl.getType();

        void dt;
    }

    onStop() {
        if (this.impl && typeof this.impl.destroy === "function") {
            try { this.impl.destroy(); } catch (_) {}
        }
        this.impl = null;
    }
}

module.exports = { PlayerCameraController };

File: assets/Scripts/player/controllers/PlayerCameraController.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local LuaClass = luaRuntime.LuaClass
local LuaConstruct = luaRuntime.LuaConstruct
PlayerCamera = require("../PlayerCamera.lua")
PlayerCameraController = LuaClass()
PlayerCameraController.name = "PlayerCameraController"
function PlayerCameraController.prototype.lua_constructor(self)
    self.ctx = nil
    self.entity = nil
    self.impl = nil
end
function PlayerCameraController.prototype.bind(self, ctx, entity)
    self.ctx = ctx
    self.entity = entity
    return self
end
function PlayerCameraController.prototype.onStart(self)
    self.impl = LuaConstruct(PlayerCamera, self.entity)
    self.impl:attach()
end
function PlayerCameraController.prototype.onUpdate(self, dt)
    local pawn = self.entity
    self.impl:update(pawn.frame)
    pawn.frame.view.yaw = self.impl:getYaw()
    pawn.frame.view.pitch = self.impl:getPitch()
    pawn.frame.view.type = self.impl:getType()
end
function PlayerCameraController.prototype.onStop(self)
    if self.impl and KTypeOf(self.impl.destroy) == "function" then
        self.impl:destroy()
    end
    self.impl = nil
end
M = {PlayerCameraController = PlayerCameraController}

return M
